core/src/main/java/org/apache/calcite/sql/fun/BodoSqlTimestampAddFunction.java

```diff
@@ -0,0 +1,372 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.sql.fun;
+
+import org.apache.calcite.avatica.util.TimeUnit;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeFactory;
+import org.apache.calcite.sql.SqlCallBinding;
+import org.apache.calcite.sql.SqlFunction;
+import org.apache.calcite.sql.SqlFunctionCategory;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.type.OperandTypes;
+import org.apache.calcite.sql.type.SqlReturnTypeInference;
+import org.apache.calcite.sql.type.SqlTypeFamily;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.calcite.tools.ValidationException;
+
+import java.util.Locale;
+
+import static org.apache.calcite.sql.validate.SqlNonNullableAccessors.getOperandLiteralValueOrThrow;
+import static org.apache.calcite.util.Static.RESOURCE;
+
+/**
+ * The <code>TIMESTAMPADD</code> function, which adds an interval to a
+ * datetime (TIMESTAMP, TIME or DATE).
+ *
+ * <p>The SQL syntax is
+ *
+ * <blockquote>
+ * <code>TIMESTAMPADD(<i>timestamp interval</i>, <i>quantity</i>,
+ * <i>datetime</i>)</code>
+ * </blockquote>
+ * Accepts both quoted and unquoted time units now.
+ *
+ * <p>Returns modified datetime.
+ */
+public class BodoSqlTimestampAddFunction extends SqlFunction {
+
+  private static final int MILLISECOND_PRECISION = 3;
+  private static final int MICROSECOND_PRECISION = 6;
+
+  private static final SqlReturnTypeInference RETURN_TYPE_INFERENCE =
+      opBinding -> {
+        final RelDataTypeFactory typeFactory = opBinding.getTypeFactory();
+        assert opBinding instanceof SqlCallBinding;
+        SqlCallBinding opBindingWithCast = (SqlCallBinding) opBinding;
+        RelDataType arg0Type = opBindingWithCast.getOperandType(0);
+        TimeUnit arg0timeUnit;
+        switch (arg0Type.getSqlTypeName()) {
+          // This must be a constant string or time unit input,
+          // due to the way that we handle the parsing
+        case CHAR:
+        case VARCHAR:
+          //This will fail if the value is a non-literal
+          try {
+            arg0timeUnit = standardizeTimeUnit("TIMESTAMPADD",
+              opBindingWithCast.getOperandLiteralValue(0, String.class),
+       opBindingWithCast.getOperandType(2).getSqlTypeName() == SqlTypeName.TIME);
+          } catch (Throwable e) {
+            throw opBindingWithCast.getValidator().newValidationError(opBindingWithCast.getCall(),
+                RESOURCE.wrongTimeUnit("TIMESTAMPADD", e.toString()));
```

Comment:
Does error.toString() just return the error message? It might return a bunch of junk too. Have you confirmed that the error message is reasonable?